trace/rate_limit.go

```diff
@@ -146,10 +145,19 @@ func (r *SharedRateLimit) startNewEpoch(epochStart time.Time) {
 		r.intervalTimer.Reset(randomOffset)
 	}
 
+	r.lock.Unlock()
+
 	// Trigger request expiration for all collectors
+	// If they never have Process() called, then their channel will fill up,
+	// so do this in a nonblocking fashion (with buffer size 1).
 	threshold := epochStart.Add(-1 * viper.GetDuration(RateLimitMaxDuration))
 	for _, child := range r.children {
-		child.epochCh <- threshold
+		select {
+		case child.epochCh <- threshold:
```

Comment:
Can `r.children` change while this executes? If so, is it safe for this to run without holding `r.lock`?

Reply:
It can change, yes.  I'll move the unlock to its original problem; this will still fix the problem.